__tests__/Auth0Client.test.ts

```diff
@@ -5,20 +5,69 @@ import unfetch from 'unfetch';
 import { verify } from '../src/jwt';
 import { MessageChannel } from 'worker_threads';
 import * as utils from '../src/utils';
-import { Auth0ClientOptions, IdToken } from '../src';
+import * as cache from '../src/cache';
+import { Auth0ClientOptions, IdToken, PopupConfigOptions } from '../src';
 import * as scope from '../src/scope';
 import { expectToHaveBeenCalledWithAuth0ClientParam } from './helpers';
 // @ts-ignore
 import { acquireLockSpy } from 'browser-tabs-lock';
+import {
+  checkSessionFn,
+  loginWithPopupFn,
+  loginWithRedirectFn,
+  setupFn,
+  TEST_ACCESS_TOKEN,
+  TEST_CLIENT_ID,
+  TEST_CODE,
+  TEST_CODE_CHALLENGE,
+  TEST_CODE_VERIFIER,
+  TEST_DOMAIN,
+  TEST_ID_TOKEN,
+  TEST_NONCE,
+  TEST_REDIRECT_URI,
+  TEST_REFRESH_TOKEN,
+  TEST_SCOPES,
+  TEST_STATE
+} from './Auth0Client.helpers';
+import { DEFAULT_POPUP_CONFIG_OPTIONS } from '../src/constants';
 
 jest.mock('unfetch');
 jest.mock('es-cookie');
 jest.mock('../src/jwt');
 jest.mock('../src/token.worker');
 
+// jest.mock('../src/utils');
+
 const mockWindow = <any>global;
 const mockFetch = (mockWindow.fetch = <jest.Mock>unfetch);
 const mockVerify = <jest.Mock>verify;
+const originalRunPopup = utils.runPopup;
+const tokenVerifier = require('../src/jwt').verify;
+
+const authorizationResponse = {
+  code: 'my_code',
+  state: TEST_STATE
+};
+
+jest
+  .spyOn(utils, 'bufferToBase64UrlEncoded')
+  .mockReturnValue(TEST_CODE_CHALLENGE);
+jest
+  .spyOn(utils, 'runPopup')
+  .mockImplementation((authorizeUrl: string, config: PopupConfigOptions) => {
+    setTimeout(() => {
+      window.dispatchEvent(
+        new MessageEvent('message', {
+          data: {
+            type: 'authorization_response',
+            response: authorizationResponse
+          }
+        })
+      );
+    }, 10);
+
+    return originalRunPopup(authorizeUrl, config);
+  });
```

Comment:
This mock is ensuring that we close a popup async after it was opened, both as an existing popup or a new popup.